notifier/processor.go

```diff
@@ -0,0 +1,231 @@
+
+	if len(added.VulnerableManifests) == 0 && len(removed.VulnerableManifests) == 0 {
+		log.Debug().Msg("0 affected manifests. will not create notifications.")
+		return nil
+	}
+
+	notifications := []Notification{}
+	create := func(r Reason, affected claircore.AffectedManifests) error {
+		for manifest, vulns := range affected.VulnerableManifests {
+			// summarize most severe vuln affecting manifest
+			// the vulns array will be sorted by most severe
+			vuln := affected.Vulnerabilities[vulns[0]]
+
+			digest, err := claircore.ParseDigest(manifest)
+			if err != nil {
+				return err
+			}
+			n := Notification{
+				Manifest: digest,
+				Reason:   r,
+			}
+			n.Vulnerability.FromVulnerability(*vuln)
+
+			notifications = append(notifications, n)
+		}
+		return nil
+	}
+	err = create(Added, added)
+	if err != nil {
+		return err
+	}
+	err = create(Removed, removed)
+	if err != nil {
+		return err
+	}
+	opts := PutOpts{
+		Updater:        e.updater,
+		UpdateID:       e.uo.Ref,
+		NotificationID: uuid.New(),
+		Notifications:  notifications,
+	}
+	err = p.store.PutNotifications(ctx, opts)
+	if err != nil {
+		return err
+	}
+	return nil
+}
+
+// safe guards against situations where creating notifications is
+// incorrect.
+//
+// if deemed safe to create notifications the previous update operation will be
+// returned
+//
+// will be performed under a distributed lock.
+func (p *Processor) safe(ctx context.Context, e Event) (bool, uuid.UUID) {
+	uoid := e.uo.Ref.String()
+	log := zerolog.Ctx(ctx).With().
+		Uint8("processor_id", p.id).
+		Str("component", "notifier/processor/Processor.Process").
+		Str("updater", e.updater).
+		Str("UOID", uoid).
+		Logger()
+	// confirm we are not making duplicate notifications
+	var errNoReceipt clairerror.ErrNoReceipt
+	_, err := p.store.ReceiptByUOID(ctx, e.uo.Ref)
+	switch {
+	case errors.As(err, &errNoReceipt):
+		// hop out of switch
+	case err != nil:
+		log.Error().Err(err).Msg("received error getting receipt by UOID")
+		return false, uuid.Nil
+	default:
+		log.Info().Msg("receipt created by another processor. will not process notifications")
+		return false, uuid.Nil
+	}
+
+	// confirm UOID is not stale and get previous UOID for diffing if exists.
+	// if no previous UOID, return false, we don't want a full diff of notifications
+	// TODO(louis) UpdateOperations signature supports getting "all" for a given updater
+	// but code path is not implemented. implement this to optimize.
+	all, err := p.matcher.UpdateOperations(ctx)
+	if err != nil {
+		log.Error().Err(err).Msg("received error getting update operations from matcher")
+		return false, uuid.Nil
+	}
+	if _, ok := all[e.updater]; !ok {
+		log.Warn().Msg("updater missing from update operations returned from matcher. matcher may have garbage collected")
+		return false, uuid.Nil
+	}
+
+	uos := all[e.updater]
+	n := len(uos)
+	if n < 2 {
```

Comment:
Are you going to change it to produce notification for dynamic updaters?

Reply:
@hdonnay @Allda can we talk about this change a bit?

Ideally, we do not want to perform the request for a very large diff, as we'd be asking the matcher to do all the work and return a large response just to say we won't process the notification.

Ideally this change happens at the matcher, and the matcher an return a `413 Payload Too Large` saying the diff is too large to process. However this still has the matcher do a least a "count" query. Maybe this is not a bad thing? 

You see that the safest thing to do is simply not process an update diff if it's the only one. I'm not held doing this, but lets discuss best way to do it moving forward.